### PR TITLE
refactor(v2): inherit from BaseLightAccount

### PR DESCRIPTION
**Pull request by jaypaik**
Introduces `BaseLightAccount`, an abstract class that different flavors of LightAccount can extend. This is hopefully minimal and flexible enough so that we don't need to overhaul this for future flavors.
- Inherits: `BaseAccount`, `TokenCallbackHandler`, `UUPSUpgradeable`, `IERC1271`
- Variables: `_1271_MAGIC_VALUE`, `_ENTRY_POINT`
- Errors: `ArrayLengthMismatch`, `NotAuthorized`
- Modifiers: `onlyOwner`
- Functions: `receive`, `execute`, `executeBatch` (both kinds), `addDeposit`, `withdrawDepositTo`, `entryPoint`, `getDeposit`, `isValidSignature` (must override), `_validateSignature` (must override), `_isFromOwner` (must override), `_isFromEntryPoint`, `_onlyOwner`, `_onlyOwnerOrEntryPoint`, `_call`, `_authorizeUpgrade`

### CODE REVIEW

File: src/common/BaseLightAccount.sol [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+    /// @dev Revert if the caller is not an owner or the account itself (when redirected through `execute`).
+    function _onlyOwner() internal view {
+        if (msg.sender != address(this) && !_isFromOwner()) {
+            revert NotAuthorized(msg.sender);
+        }
+    }
+
+    /// @dev Require that the call is from the entry point or an owner.
+    function _onlyOwnerOrEntryPoint() internal view {
```
Comment (adam-alchemy) [inline src/common/BaseLightAccount.sol:113]: I like this modifier setup

File: src/common/BaseLightAccount.sol [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        if (!success) {
+            assembly {
+                revert(add(result, 32), mload(result))
+            }
+        }
+    }
+
+    function _authorizeUpgrade(address newImplementation) internal view override {
+        (newImplementation);
+        _onlyOwner();
```
Comment (adam-alchemy) [inline src/common/BaseLightAccount.sol:130]: Could we make this use `_onlyOwnerOrEntryPoint`?

**Review by adam-alchemy — APPROVED** [review]

File: src/common/BaseLightAccount.sol [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    /// @inheritdoc IERC1271
+    /// @dev Must override to support ERC-1271 signature validation.
+    function isValidSignature(bytes32 hash, bytes memory signature) public view virtual override returns (bytes4);
+
+    /// @dev Must override to support user op validation.
+    function _validateSignature(PackedUserOperation calldata userOp, bytes32 userOpHash)
+        internal
+        virtual
+        override
+        returns (uint256);
```
Comment (adam-alchemy) [inline src/common/BaseLightAccount.sol]: Do we get this unimplemented abstract function from `BaseAccount` already?

Comment (jaypaik) [reply src/common/BaseLightAccount.sol]: Yup, nice catch

File: src/common/BaseLightAccount.sol (hunk 2 above)
Comment (jaypaik) [reply src/common/BaseLightAccount.sol:130]: Yup, I'll address the modifier changes in another PR.

**Pull request merged**